### PR TITLE
Fully remove submodule handling from bootstrap.py

### LINKED ISSUE #90764: `x.py test` try to access network when the compiler builds using vendored crates
We are building the compiler and tools using the vendored crates properly (via `x.py build`) in systems with no network access. But when running the tests (`x.py test`) it tries to access the network.

Setting properly the envs vars that x.py set, this can be reproduced with:

```
# /home/abuild/rpmbuild/BUILD/rustc-1.56.1-src/build/bootstrap/debug/bootstrap test

Building stage0 tool tidy (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.24s
tidy check
Checking which error codes lack tests...
* 627 error codes
* highest error code: E0785
Found 501 error codes
Found 0 error codes with no tests
Done!
thread '<unnamed>' panicked at 'cmd.exec() failed with Error during execution of `cargo metadata`:     Updating git repository `https://github.com/bjorn3/rust-ar.git`
warning: spurious network error (2 tries remaining): failed to resolve address for github.com: Temporary failure in name resolution; class=Net (12)
warning: spurious network error (1 tries remaining): failed to resolve address for github.com: Temporary failure in name resolution; class=Net (12)
error: failed to get `ar` as a dependency of package `rustc_codegen_cranelift v0.1.0 (/home/abuild/rpmbuild/BUILD/rustc-1.56.1-src/compiler/rustc_codegen_cranelift)`

Caused by:
  failed to load source for dependency `ar`

Caused by:
  Unable to update https://github.com/bjorn3/rust-ar.git?branch=do_not_remove_cg_clif_ranlib#de9ab0e5

Caused by:
  failed to fetch into: /home/abuild/.cargo/git/db/rust-ar-9b35aff8ad678e06

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  failed to resolve address for github.com: Temporary failure in name resolution; class=Net (12)
', src/tools/tidy/src/deps.rs:293:20
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any { .. }', src/tools/tidy/src/main.rs:77:9


command did not execute successfully: "/home/abuild/rpmbuild/BUILD/rustc-1.56.1-src/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/home/abuild/rpmbuild/BUILD/rustc-1.56.1-src" "/home/abuild/rpmbuild/BUILD/rust-1.56.1-x86_64-unknown-linux-gnu/usr/bin/cargo" "/home/abuild/rpmbuild/BUILD/rustc-1.56.1-src/build" "8"
expected success, got: exit status: 101
```

Seems that the call is be present in `rust-tidy` (in `deps.rs`). IIUC should use the `--offline` parameter, and I do not think that is checking the presence of `.cargo/config` per the `ar` error (that is present in the `vendor` directory)

### PR DESCRIPTION
**Pull request by jyn514**
**NOTE**: If you are coming here because of the link in the changelog, the new way to vendor the workspace is by running `cargo vendor --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml` manually. If you want to use the pre-downloaded cargo, run `make prepare` first, then use the cargo in `build/$TARGET/stage0/bin/cargo`. Unlike the rest of bootstrapping, this is mostly independent of the version and should work even with fairly old versions of cargo.

---

These submodules were previously updated in python because Cargo gives a hard error if toml files
are missing from the workspace:

```
error: failed to load manifest for workspace member `/home/jnelson/rust-lang/rust/src/tools/rls`

Caused by:
  failed to read `/home/jnelson/rust-lang/rust/src/tools/rls/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
failed to run: /home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /home/jnelson/rust-lang/rust/src/bootstrap/Cargo.toml
```

However, bootstrap doesn't actually need to be part of the workspace.
Remove it so we can move submodule handling fully to Rust, avoiding duplicate code between Rust and Python.

Note that this does break `cargo run`; it has to be `cd src/bootstrap && cargo run` now.
Given that we're planning to make the main entrypoint a shell script (or rust binary),
I think this is a good tradeoff for reduced complexity in bootstrap.py.

To get this working, I also had to remove support for vendoring when using the git sources, because `cargo vendor` requires all submodules to be checked out. I think this is ok; people who care about this are likely already using the pre-vendored `rustc-src` tarball.

Fixes https://github.com/rust-lang/rust/issues/90764. Helps with #94829

### CODE REVIEW

File: src/bootstrap/Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,459 @@
+# This file is automatically @generated by Cargo.
```
Comment (bjorn3) [inline src/bootstrap/Cargo.lock:1]: Please add `src/bootstrap/Cargo.lock` to the list of `Cargo.lock` files tidy looks at.

**Comment by ehuss** [comment]
Won't this cause problems for vendoring?

**Comment by bjorn3** [comment]
Probably. `src/bootstrap` would need to be added to the list of crates to vendor dependencies for at https://github.com/rust-lang/rust/blob/396e63dbcaa085d7d5f9491ea95bd2f839dfbbe8/src/bootstrap/bootstrap.py#L983 and it would still need the submodule updates before performing the vendoring. Even right now rust-analyzer is checked out by bootstrap.py when vendoring, despite it being handled by rustbuild when not vendoring.

**Comment by jyn514** [comment]
@bjorn3 I added bootstrap to the tidy check, but I think we might not actually need to be so restrictive with bootstrap as with the other tools? We don't distribute bootstrap, it's only used in-tree, so I don't think it adds additional obligations to users building from source. @Mark-Simulacrum or someone else from core might know more.

> it would still need the submodule updates before performing the vendoring

Hmm, good point. Can we only vendor rustbuild itself in bootstrap.py, and add the rest of the vendoring in rustbuild? Merging the .cargo/config files seems kind of tricky, though ...

**Comment by jyn514** [comment]
Btw, why is vendoring handled through x.py? Why don't we ask people to configure vendoring themselves if they care?

**Comment by bjorn3** [comment]
> Btw, why is vendoring handled through x.py? Why don't we ask people to configure vendoring themselves if they care?

The source tarballs are unconditionally vendored.

**Comment by jyn514** [comment]
> it would still need the submodule updates before performing the vendoring

If we need to maintain the submodule code in bootstrap.py, it kind of defeats the point of this change :/ I wonder if we can somehow get git to *only* fetch the Cargo.toml / Cargo.lock files that we need? Or move vendoring to rustbuild somehow?

**Comment by jyn514** [comment]
I'm going to move this to waiting-on-review since I'm not sure the right approach to take for vendoring when submodules haven't yet been cloned.

@rustbot ready

**Comment by jyn514** [comment]
> Or move vendoring to rustbuild somehow?

Ok, so here's a terrible idea: have multiple vendor directories, one only for bootstrap and nothing else (to avoid pulling in crates that are submodules in the workspace), and one for the workspace. Bootstrap.py will manage the first, rustbuild will manage the second. This has the upside of letting us get rid of submodule handling; it has the downside that now we have to duplicate all the vendoring code in rustbuild. That's probably not *too* terribly hard; `cargo vendor` is a lot simpler than `git submodule`.

**Comment by Mark-Simulacrum** [comment]
It looks like we're trying to support vendoring sources in x.py, but I'm not actually sure that's necessary. Supporting *running* with vendored sources makes sense to me (i.e., if you downloaded a src tarball, we shouldn't need further network from there). But if you want to produce such a tarball, then I think it's either on you to use x.py dist rust-src... I'm not sure there's a strong use case for enabling vendoring when working locally and having x.py vendor just before the build.

I'd be OK saying that if folks want that they can manually find the command and run it, which removes the need to support vendoring from x.py, I think; it just leaves the need to support *using* a vendor directory, which is much simpler (and doesn't need duplication).

@cuviper -- is my impression that we don't need to support vendoring from a regular checkout correct? I don't know what would push someone to want that.

**Comment by cuviper** [comment]
I can imagine that someone building from git might want to init submodules and vendor, then work offline from there. Maybe that's someone getting on an airplane -- but as long as they can pre-fetch, that kind of use doesn't really need vendoring. If you really need a fully self-contained source dir, the dist command seems fine.

For my Fedora/RHEL builds, I'm always using the source tarball.

**Comment by jyn514** [comment]
> it just leaves the need to support _using_ a vendor directory, which is much simpler (and doesn't need duplication).

btw this isn't possible today, because bootstrap forcibly removes your config: https://github.com/rust-lang/rust/blob/dc80ca78b6ec2b6bba02560470347433bcd0bb3c/src/bootstrap/bootstrap.py#L960-L961

I'm going to remove this as part of this change, but I do worry this is stacking a lot of changes on top of each other ...

**Comment by cuviper** [comment]
> btw this isn't possible today, because bootstrap forcibly removes your config:

But just above that is the code that creates a config for vendoring.

**Comment by jyn514** [comment]
@cuviper so, the logic here is a little convoluted - here's the whole thing: https://github.com/rust-lang/rust/blob/dc80ca78b6ec2b6bba02560470347433bcd0bb3c/src/bootstrap/bootstrap.py#L942-L961

in psuedo code:
```
if use_vendoring:
  if .cargo/config exists:
    print a warning and continue
  else:
    create .cargo/config
else:
  remove .cargo/config
```
After this change, `self.use_vendored_sources` will never be true, so it will always forcibly remove your config unless I delete that snippet.

**Comment by jyn514** [comment]
In other words, bootstrap.py currently *only* supports vendoring that's managed by bootstrap, it doesn't allow you to manage your own vendoring.

**Comment by cuviper** [comment]
I don't care about vendoring it myself, but I do care that I can use the vendored sources from the dist tarball.

**Comment by jyn514** [comment]
> I don't care about vendoring it myself, but I do care that I can use the vendored sources from the dist tarball.

That's kind of my point, yeah, this change is necessary to keep allowing that.

How do you do this currently, do you set `use_vendored_sources` in config.toml? the .tar.gz doesn't come with a .cargo/ directory.

**Comment by cuviper** [comment]
I use `./configure ... --enable-vendor`, which I believe sets `build.vendor` in the resulting `config.toml`.

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -1052,12 +926,10 @@ def bootstrap(help_triggered):
     if not os.path.exists(build.build_dir):
         os.makedirs(build.build_dir)
     lock = acquire_lock(build.build_dir)
-    build.update_submodules()
 
     # Fetch/build the bootstrap
     build.download_toolchain()
     sys.stdout.flush()
-    build.ensure_vendored()
     build.build_bootstrap()
```
Comment (jyn514) [inline src/bootstrap/bootstrap.py:933]: Love how there's basically nothing left :) I think just the build lock is left and then we can start using rustbuild as the entrypoint

**Comment by semarie** [comment]
I am having problems with using vendoring crates and downloaded tarball (currently with nighty and beta), and I am wondering if the commit could be linked to the problem.

see #100364 for details.

**Pull request merged**